tests/unit/test_rose_stem_units.py

```diff
@@ -54,3 +74,220 @@ def test_get_source_opt_from_args(tmp_path, monkeypatch, args, expect):
+                'url': 'Irrelevent, it\'s mocked away, but required.'
+            },
+            (
+                0,
+                (
+                    "location{primary}[mortimer] = "
+                    "svn://subversive/rogermortimer\n"
+                    "location{primary}[fenwick] = "
+                    "svn://subversive/johnfenwick\n"
+                    "location{primary}[waltheof] = "
+                    "svn://subversive/waltheof\n"
+                ),
+            ),
+            'waltheof',
+            id='all paths true'
+        ),
+        param(
+            {
+                'root': 'svn://subversive',
+                'project': 'waltheof',
+                'url': 'Irrelevent, it\'s mocked away, but required.'
+            },
+            (0, "location{primary} = svn://subversive/waltheof\n"),
+            None,
+            id='no kp result'
+        )
+    ]
+)
+def test__get_project_from_url(
+    get_StemRunner, source_dict, mockreturn, expect
+):
+    stemrunner = get_StemRunner({'popen': MockPopen(mockreturn)})
+    project = stemrunner._get_project_from_url(source_dict)
+    assert project == expect
+
+
+@pytest.mark.parametrize(
+    'source, expect',
+    (
+        (None, 'cwd'),
+        ('foo/bar', 'some_dir'),
+    )
+)
+def test__generate_name(get_StemRunner, monkeypatch, tmp_path, source, expect):
+    """It generates a name if StemRunner._ascertain_project fails.
+
+    (This happens if the workflow source is not controlled with FCM)
+    """
+    monkeypatch.chdir(tmp_path)
+
+    # Case: we've set source:
+    source = (tmp_path / source / expect) if expect == 'some_dir' else None
+    # Case: we've not set source:
+    expect = tmp_path.name if expect == 'cwd' else expect
+
+    stemrunner = get_StemRunner({}, {'source': source})
+    assert stemrunner._generate_name() == expect
+
+
+@pytest.mark.parametrize(
+    'stem_sources, expect',
+    (
+        ('given', True),
+        ('given', False),
+        ('infer', True),
+        ('infer', False),
+    )
+)
+def test__this_suite(
+    get_StemRunner, monkeypatch, tmp_path, stem_sources, expect
+):
+    """It returns a sensible suite-dir."""
+    stem_suite_subdir = tmp_path / 'rose-stem'
+    stem_suite_subdir.mkdir()
+
+    if stem_sources == 'infer':
+        stem_sources = []
+        monkeypatch.setattr(
+            cylc.rose.stem.StemRunner,
+            '_ascertain_project',
+            lambda x, y: [0, str(tmp_path)]
+        )
+    else:
+        stem_sources = [tmp_path]
+
+    if expect:
+        (stem_suite_subdir / 'rose-suite.conf').write_text(
+            'ROSE_STEM_VERSION=1')
+        stemrunner = get_StemRunner({}, {'stem_sources': stem_sources})
+        assert stemrunner._this_suite() == str(stem_suite_subdir)
+    else:
+        with pytest.raises(RoseSuiteConfNotFoundException):
+            stemrunner = get_StemRunner({}, {'stem_sources': stem_sources})
+            stemrunner._this_suite()
```

Comment:
If the exception is expected on the second line, then suggest moving the first out of there
```suggestion
        stemrunner = get_StemRunner({}, {'stem_sources': stem_sources})
        with pytest.raises(RoseSuiteConfNotFoundException):
            stemrunner._this_suite()
```